### PR TITLE
JDK-8325898: ChoiceFormat returns erroneous result when formatting bad pattern

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
Please review this PR which handles an edge case pattern bug with ChoiceFormat.

```
var d = new ChoiceFormat("0#foo|1#bar|baz|") // creates cFmt equivalent to "0.0#foo|1.0#bar|1.0#"
d.format(1) // unexpectedly returns ""
```

Not only does this lead to faulty formatting results, but breaks the ChoiceFormat class invariant of duplicate limits.

It would have been preferable if either an exception was thrown when the ChoiceFormat was initially created, or the ChoiceFormat formatting 1 returned a value of "bar".

For comparison,
```
var d = new ChoiceFormat("0#foo|1#bar|baz") // creates cFmt equivalent to "0.0#foo|1.0#bar"
d.format(1) // returns "bar"
```

After this change, the first code snippet now returns "bar" when formatting 1 and discards the "baz" as the second code snippet does.

The fix prevents a limit/format entry from being built when the current parsing mode has not yet processed the limit and relation. The previous implementation would build an additional limit/format of 1 with an empty string. The `break` allows for the discarding of the incorrect portion and continuing. Alternatively an exception could have been thrown. For consistency with the second code snippet, the former was picked.

https://github.com/openjdk/jdk/pull/17856 may be of interest, which is a semi-related specification fix.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325898](https://bugs.openjdk.org/browse/JDK-8325898): ChoiceFormat returns erroneous result when formatting bad pattern (**Bug** - P4)

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to [de38a988](https://git.openjdk.org/jdk/pull/17883/files/de38a9881660ca8ab52ebfa3da67b090d4585df0)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17883/head:pull/17883` \
`$ git checkout pull/17883`

Update a local copy of the PR: \
`$ git checkout pull/17883` \
`$ git pull https://git.openjdk.org/jdk.git pull/17883/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17883`

View PR using the GUI difftool: \
`$ git pr show -t 17883`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17883.diff">https://git.openjdk.org/jdk/pull/17883.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17883#issuecomment-1947151792)

### CODE REVIEW

File: src/java.base/share/classes/java/text/ChoiceFormat.java [hunk 1]
```diff
@@ -249,86 +251,110 @@ public void applyPattern(String newPattern) {
...
-        for (int i = 0; i < segments.length; ++i) {
-            segments[i] = new StringBuilder();
-        }
-        double[] newChoiceLimits = new double[30];
-        String[] newChoiceFormats = new String[30];
-        int count = 0;
-        int part = 0; // 0 denotes limit, 1 denotes format
-        double startValue = 0;
-        double oldStartValue = Double.NaN;
+        Objects.requireNonNull(newPattern, "newPattern must not be null");
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/ChoiceFormat.java:254]: I think this null check should be at the public method level, not the internal implementation. This technically ends up revealing the internal variable name (although it is the same with the public argument names as of now)

File: src/java.base/share/classes/java/text/ChoiceFormat.java [hunk 2]
```diff
@@ -249,86 +251,110 @@ public void applyPattern(String newPattern) {
...
-                newChoiceFormats[count] = segments[1].toString();
-                ++count;
-                oldStartValue = startValue;
-                segments[1].setLength(0);
-                part = 0;
-            } else {
-                segments[part].append(ch);
-            }
+    // Used to explicitly define the segment mode while applying a pattern
+    private enum Segment {
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/ChoiceFormat.java:332]: Do we need a new enum? Would introducing a new enum solve something that cannot be done with the existing implementation?

File: src/java.base/share/classes/java/text/ChoiceFormat.java [hunk 3]
```diff
@@ -249,86 +251,110 @@ public void applyPattern(String newPattern) {
...
-                segments[part].append(ch);
-            } else if (part == 0 && (ch == '<' || ch == '#' || ch == '\u2264')) {
-                // Only consider relational symbols if parsing the limit segment (part == 0).
-                // Don't treat a relational symbol as syntactically significant
-                // when parsing Format segment (part == 1)
-                if (segments[0].length() == 0) {
-                    throw new IllegalArgumentException("Each interval must"
-                            + " contain a number before a format");
-                }
+            switch(ch) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/text/ChoiceFormat.java:268]:
```suggestion
            switch (ch) {
```

File: src/java.base/share/classes/java/text/ChoiceFormat.java (hunk 2 above)
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/text/ChoiceFormat.java:332]: No, we don't _need_ a new enum, although I prefer it for readability.

For example, I think the following is more clear:
- `Segment.LIMIT.bldr.toString()`over `segments[1].toString();`.
- `seg = Segment.LIMIT;` over `part = 0;`

Can also no longer do something such as `part = 2;` even if unlikely.

Although I can see the argument of not wanting an enum as `part` is either only `0` or `1`. Either is OK with me.

Comment (naotoj) [reply src/java.base/share/classes/java/text/ChoiceFormat.java:331]: OK, now the new enum can be shared with multiple threads with this implementation, but I guess it is OK as ChoiceFormat does not guarantee synchronization.

File: src/java.base/share/classes/java/text/ChoiceFormat.java [hunk 4]
```diff
@@ -249,86 +252,108 @@ public void applyPattern(String newPattern) {
...
-                part = 0;
-            } else {
-                segments[part].append(ch);
-            }
+    // Used to explicitly define the segment mode while applying a pattern
+    private enum Segment {
+        LIMIT(new StringBuilder()),
+        FORMAT(new StringBuilder());
+
+        private final StringBuilder bldr;
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/ChoiceFormat.java:335]: Maybe needs a more meaningful name than `bldr`? At first look, I thought it was intended for building a segment.

File: src/java.base/share/classes/java/text/ChoiceFormat.java [hunk 5]
```diff
@@ -249,86 +252,95 @@ public void applyPattern(String newPattern) {
...
+     */
+    private static double stringToNum(String str) {
+        double num;
+        if (str.equals("\u221E")) {
+            num = Double.POSITIVE_INFINITY;
+        } else if (str.equals("-\u221E")) {
+            num = Double.NEGATIVE_INFINITY;
+        } else {
+            num = Double.parseDouble(str);
+        }
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/ChoiceFormat.java:342]: Could use enhanced switch?

**Review by naotoj — APPROVED** [review]
LGTM, with a minor suggestion.

File: src/java.base/share/classes/java/text/ChoiceFormat.java (hunk 5 above)
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/text/ChoiceFormat.java]: Yes, much cleaner that way.

**Review by naotoj — APPROVED** [review]

**Pull request closed without merging**